src/caslist/mod.rs

```diff
@@ -0,0 +1,103 @@
+use std::marker::PhantomData;
+use std::iter::IntoIterator;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+#[cfg(test)]
+mod test;
+
+/// A simple linked list, based on compare-and-swap. Values can only
+/// be prepended to the front. At present, they can never be removed.
+pub struct CasList<T> {
+    head: AtomicPtr<Cell<T>>, // if not null, a unique, transmuted ptr to a Box<T>
+}
+
+struct Cell<T> {
+    next: *mut Cell<T>, // if not null, a unique, transmuted ptr to a Box<T>
+    data: T,
+}
+
+impl<T> CasList<T> {
+    pub fn new() -> Self {
+        CasList {
+            head: AtomicPtr::new(ptr::null_mut())
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        let head = self.head.load(Ordering::Acquire);
+        head.is_null()
+    }
+
+    pub fn prepend(&self, value: T) {
+        unsafe {
+            let cell = Box::new(Cell {
+                next: ptr::null_mut(),
+                data: value
+            });
+
+            // We're going to transfer ownership of this box into the
+            // list. Note that none of this code below can panic,
+            // which is important for avoiding leaks!
+            let cell: *mut Cell<T> = mem::transmute(cell);
+            loop {
+                let head = self.head.load(Ordering::Relaxed);
+                (*cell).next = head;
+                if self.head.compare_and_swap(head, cell, Ordering::Release) == head {
+                    return;
+                }
+            }
+        }
+    }
+
+    pub fn iter(&self) -> Iter<T> {
+        Iter {
+            phantom: PhantomData,
+            ptr: self.head.load(Ordering::Acquire),
+        }
+    }
+}
+
+impl<T> Drop for CasList<T> {
+    fn drop(&mut self) {
+        unsafe {
+            let mut ptr: *mut Cell<T> = self.head.load(Ordering::Acquire);
```

Comment:
This ordering makes me uneasy - I'm afraid it's not sufficient and might potentially load a slightly older head than the actual current head. In other words, some of the newest cells might not get destructed and thus memory will be leaked.

The `Acquire` ordering will merely load *some head* from the past and will ensure that we see all writes to memory that happened before *that head* was installed. However, it is not guaranteed that the head that we just loaded really is the most up-to-date head.

Instead, write a usual CAS loop to set the head back to null. After that, traverse the list and destruct all cells. By CAS-ing the head to null we make sure that we're actually holding the latest head and "drain" the list.
The CAS loop might look like this:
```rust
let mut head = self.head.load(Ordering::Acquire);
loop {
    let previous = self.head.compare_and_swap(head, ptr::null_mut(), Ordering::AcqRel);
    if previous == head {
        break;
    } else {
        head = previous;
    }
}
// Now Traverse and destruct the list starting from `head`...
```

Notice the `Acquire` and `AcqRel` orderings - we need to make sure that we see all writes to memory that happened before the loaded head was installed.